### DISCUSSION TITLE
VectorVariable Gradient at a Point

### DISCUSSION THREAD

**Discussion by maxnezdyur** [idaholab/moose#23478] [Q&A General] [answered]
Is there a way to get a VectorVariable gradient at a point? Just looking for a function that I can call within the VectorNodalBC class.

**Comment by ngrilli** [comment]
@maxnezdyur  I have done something similar here:

https://github.com/ngrilli/c_pfor_am/blob/main/src/auxkernels/ArrayDirectionalDerivative.C

In my case it is a directional derivative, but you can simply use the gradient,
the variable that express the gradient is:

_grad_variable[_qp](j*_var.count()+i)

Trust this helps,
Nicolò

**Comment by GiudGiud** [comment, marked as answer]
are you thinking of coupledGradient from the coupleable interface ?

**Reply by maxnezdyur** [reply]
So when using the VectorNodalBC class, I need to find the gradient at a location within the element. Do you know how coupledGradient works when you are on a node? I may be overthinking this and it would work. I finally found this in system.h, it has a few functions for `point_gradient`, but I don't know if it works for vector variables.

**Reply by GiudGiud** [reply]
so coupledVectorGradient then for vector variables most likely. 

and this is a nodal vector variable right? a Lagrange_vec ?

**Reply by maxnezdyur** [reply]
Yes this is Lagrange_vec. 
So `coupledVectorGradient` calls `var->gradSln()` which calls `_element_data->gradSln(Moose::Current)`. I may be misunderstanding this but when using the VectorNodalBC, which element gradient will it be giving? A node may be connected to multiple elements with different gradients.

**Reply by GiudGiud** [reply]
yep this doesnt look right.

The coupleable interface has nodal value routines. (coupledNodalValue for example)

I think you may need to build the gradient ones because I dont see them

**Reply by maxnezdyur** [reply]
I probably will try to build version of `point_gradient` that can work with Lagrange_vec.

**Reply by GiudGiud** [reply]
For a Lagrange variable, the derivative is not continuous on nodes is it?

This is typically why these routines are missing. For other higher order types we could have been better defined

**Reply by maxnezdyur** [reply]
Yes I agree. That's why I am trying to find the gradient at a point that I know is within an element so it should be better defined.

**Reply by lindsayad** [reply]
I do think that `point_gradient` is the best (maybe only) API to use here. If it doesn't work yet for vector finite element types ... actually yea as I'm typing this I realize that it most certainly won't work. Let me think a little bit more

**Reply by lindsayad** [reply]
Ok here is some example code
```c++
auto fe = FEGenericBase<RealVectorValue>::build(elem->dim(), _var.feType());
const auto & dphi = fe->get_dphi();
const std::vector<Point> my_points = {point};
fe->reinit(elem, &my_points);
std::vector<dof_id_type> dof_indices;
const auto & dof_map = _sys.dofMap();
dof_map.dof_indices(elem, dof_indices, _var.number());
mooseAssert(dof_indices.size() == dphi.size(), "These must match");
RealTensorValue grad = 0;
for (const auto i : index_range(dof_indices))
  grad += _sys.solution()(dof_indices[i]) * dphi[i][0];
```
@roystgnr might have some other suggestions. There are definitely options out of `FEMContext` but a lot of those look pretty heavy (at least in terms of line count ... maybe not in terms of performance)

**Reply by roystgnr** [reply]
`FEMContext` is a good way to go if you already have one in use for a quadrature loop; I don't think it offers anything over the lower-level APIs if you just need a one-off.
